rfcs/0051-hard-fork-package-generation.md

```diff
@@ -0,0 +1,148 @@
+
+__IMPORTANT NOTE:__ dumping ledgers back to back needs to not break a node (this may require additional work).
+
+### Capturing Genesis State for Unplanned Hard Forks
+
+In the case of an unplanned (emergency) hard fork, we do not have the liberty of allowing the daemon to automatically dump the state that we will start the new chain from. Additionally, the state we want to take could be arbitrarily far back in history (obviously we won't take something that is too far back in history since we don't want to undo much history, but we can't make assumptions about how far back we need to look).
+
+In such a circumstance, we must dump this state from the archive db rather than from an actively running daemon. The replayer tool is capable of materializing staged ledgers, snarked ledgers, and epoch ledgers for arbitrary blocks by replaying transactions from the archie db. It is important that this tool should take at most a couple of hours to execute, as the time spent executing this tool is a further delay in generating a new hard fork package for the emergency release.
+
+### Generating the Hard Fork Package
+
+The hard fork package will be generated in a buildkite pipeline. The buildkite pipeline will accept environment variables providing URLs to download the unmigrated genesis state from. Because different jobs require different parts of the genesis state to run, we will split up the unmigrated genesis state inputs into the following files:
+
+- `block.json` (the full protocol state of the dumped block)
+- `staged_ledger.json` (the contents of the staged ledger of the dumped block)
+- `next_staking_ledger.json` (the contents of the next staking ledger of the dumped block)
+- `staking_ledger.json` (the contents of the staking ledger of the dumped block)
+
+Here is a diagram of the various buildkite jobs required to generate the hard fork package:
+
+![](res/hard-fork-package-generation-buildkite-pipeline.dot.png)
+
+The `build_daemon` job will build the daemon with the correct compile time configuration for mainnet (`mainnet.mlh`). The hard fork specific configuration will be provided via a static runtime configuration that is bundled with the hard fork release.
+
+The `generate_network_config` job will generate the runtime configuration for the hard fork network, embedding the correct ledger hashes and correct fields for `fork_previous_length`, `fork_previous_state_hash` and `fork_previous_global_slot`.
+
+* IMPORTANT: when does the genesis\_timestamp get written in? has is it configured?
```

Comment:
Shouldn't it get its own json file, which gets folded into the runtime config above?

Reply:
I was thinking something like that, yes. It seems that we should have a target genesis timestamp for the new network available at the top of the pipeline, which we can pass in as an env var or in some json data. What I was uncertain of is whether we need to be able to "repackage" with a different genesis timestamp, if we miss that window. The pipeline may be able to support this, where we reuse the other artifacts we already produced and just repackage the runtime config, but I'm not sure if there is a need that warrants the complexity.

Reply:
@nholland94 is this genesis timestamp included in chain id computation? If so, having another chain id changes anything or it's ok?